pkg/issuer/vault/issue.go

```diff
@@ -0,0 +1,200 @@
+}
+
+// signCertificate returns a signed x509.Certificate object for the given
+// *v1alpha1.Certificate crt.
+func (v *Vault) signCertificate(crt *v1alpha1.Certificate, key *rsa.PrivateKey) ([]byte, error) {
+	commonName, err := pki.CommonNameForCertificate(crt)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting Common Name from certificate: %s", err.Error())
+	}
+
+	altNames, err := pki.DNSNamesForCertificate(crt)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting DNS Names from certificate: %s", err.Error())
+	}
+
+	if len(commonName) == 0 && len(altNames) > 0 {
+		commonName = altNames[0]
+	}
+
+	template := pki.GenerateCSR(commonName, altNames...)
+
+	derBytes, err := x509.CreateCertificateRequest(rand.Reader, template, key)
+	if err != nil {
+		return nil, fmt.Errorf("error creating x509 certificate: %s", err.Error())
+	}
+
+	pemRequestBuf := &bytes.Buffer{}
+	err = pem.Encode(pemRequestBuf, &pem.Block{Type: "CERTIFICATE REQUEST", Bytes: derBytes})
+	if err != nil {
+		return nil, fmt.Errorf("error encoding certificate request: %s", err.Error())
+	}
+
+	return v.requestVaultCert(commonName, altNames, pemRequestBuf.String())
+}
+
+func (v *Vault) requestVaultCert(commonName string, altNames []string, csr string) ([]byte, error) {
+	client, err := vault.NewClient(nil)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing Vault client: %s", err.Error())
+	}
+
+	client.SetAddress(v.issuer.GetSpec().Vault.Server)
+
+	tokenRef := v.issuer.GetSpec().Vault.Auth.TokenSecretRef
+	token, err := vaultTokenRef(v.secretsLister, v.issuerResourcesNamespace, tokenRef.Name, "token")
+	if err != nil {
+		return nil, fmt.Errorf("error reading Vault token from secret %s/%s: %s", v.issuerResourcesNamespace, tokenRef.Name, err.Error())
+	}
+
+	client.SetToken(token)
+
+	glog.V(4).Infof("Vault certificate request for commonName %s altNames: %q", commonName, altNames)
+
+	parameters := map[string]string{
+		"common_name": commonName,
+		"alt_names":   strings.Join(altNames, ","),
+		"ttl":         time.Duration(certificateDuration).String(),
+		"csr":         csr,
+		"exclude_cn_from_sans": "true",
+	}
+
+	url := fmt.Sprintf("/v1%s", v.issuer.GetSpec().Vault.Path)
+
+	request := client.NewRequest("POST", url)
+
+	err = request.SetJSONBody(parameters)
+	if err != nil {
+		return nil, fmt.Errorf("error encoding Vault parameters: %s", err.Error())
+	}
+
+	resp, err := client.RawRequest(request)
+	if err != nil {
+		return nil, fmt.Errorf("error calling Vault server: %s", err.Error())
+	}
+
+	defer resp.Body.Close()
+
+	vaultResult := VaultResult{}
+	resp.DecodeJSON(&vaultResult)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode JSON payload: %s", err.Error())
+	}
+
+	certBytes := &bytes.Buffer{}
+	certBytes.WriteString(vaultResult.Data.Cert)
+	certBytes.WriteString("\n")
+	for _, v := range vaultResult.Data.Chain {
+		certBytes.WriteString(v)
+		certBytes.WriteString("\n")
+	}
```

Comment:
This is okay I think, but it's worth noting right now we don't expose any way to obtain the CA that signed a Certificate via cert-manager (this is something that would be great to change, but I've not been able to find a way to get a CA certificate from an ACME server (cc @cpu)).

iirc, the certificate chain here doesn't actually include that CA certificate anyway so this should be okay to include here.

Reply:
> I've not been able to find a way to get a CA certificate from an ACME server (cc @cpu)).

@munnerz For the V1 API the `new-cert` HTTP response includes [a `Link` header with an `up` relation](https://github.com/letsencrypt/boulder/blob/c0ffa3d5d1ab389614ffb3094de69798abc3af72/wfe/wfe.go#L1541) that points to where you can find the issuer certificate for the issued leaf certificate.

For the V2 API things are much easier. After finalizing an order the URL of the certificate in the order object returns the full PEM encoded certificate chain. When you GET the cert URL you'll get both the leaf and the intermediate(s).

Hope that helps!